Add Allocator support to executors

upsj opened the pull request:
This adds support for custom allocators to our executors (except for DPC++). We design follows the `malloc/free` function pair, with base types for all executor memory spaces to avoid users accidentally using incompatible Executor/Allocator pairs. I also used this opportunity to clean up the executor.* files and separate forward-declarations, streams, allocators and executors into different files, and deprecate the device_reset flag, which is no longer necessary for tests.

#### Why model `malloc/free` and not more allocator-friendly interfaces?

Without an interface-break, we cannot change the `raw_alloc/raw_free` interface, so we need to stick with the minimum that can be callsed from `raw_alloc(size)` and `raw_free(pointer)`. Additionally,  none of the allocator wrappers I implemented can profit from any additional size information or even an opaque allocator info struct.

MarcelKoch commented:
Since there were also other design options for the allocator interface, would you mind giving a very brief explanation, why you chose this approach over the others?

upsj commented:
@MarcelKoch Good point, I added an explanation - basically interface stability. Also none of the allocator wrappers I implemented can profit from any additional information.

yhmtsai commented on hip/base/executor.hip.cpp:
```diff
@@ -60,32 +56,22 @@ namespace gko {
-            }
-        });
+        new HipExecutor(device_id, std::move(master),
+                        std::make_shared<HipAllocator>(), stream));
```
also create_from_mode?

upsj replied:
done

upsj commented on include/ginkgo/core/base/executor.hpp:
```diff
@@ -1355,26 +1307,14 @@ class EnableDeviceReset {
      *
      * @param device_reset  whether to allow a device reset or not
      */
-    void set_device_reset(bool device_reset) { device_reset_ = device_reset; }
+    void set_device_reset(bool device_reset) {}
 
     /**
      * Returns the current status of the device reset boolean for this executor.
      *
      * @return the current status of the device reset boolean for this executor.
      */
-    bool get_device_reset() { return device_reset_; }
```
true, we can do without this change, only return `false` every time instead

upsj commented on include/ginkgo/core/base/memory.hpp:
```diff
@@ -0,0 +1,215 @@
+
+    virtual void* allocate(size_type num_bytes) const = 0;
+
+    virtual void deallocate(void* ptr) const = 0;
```
this is using the nomenclature from https://en.cppreference.com/w/cpp/memory/allocator

upsj commented on test/utils/executor.hpp:
```diff
@@ -146,7 +149,13 @@ class CommonTestFixture : public ::testing::Test {
+        :
+#if defined(GKO_TEST_NONDEFAULT_STREAM) && \
+    (defined(GKO_COMPILING_CUDA) || defined(GKO_COMPILING_HIP))
+          stream{0},
```
no, it's a newly created stream on the device 0. `cuda_stream` can't be constructed from a `cudaStream_t` because of the unknown ownership.

yhmtsai replied:
I see

upsj commented on cuda/base/executor.cpp:
```diff
@@ -123,41 +131,14 @@ void OmpExecutor::raw_copy_to(const CudaExecutor* dest, size_type num_bytes,
 void* CudaExecutor::raw_alloc(size_type num_bytes) const
 {
-    void* dev_ptr = nullptr;
     detail::cuda_scoped_device_id_guard g(this->get_device_id());
```
I don't see extremely strong arguments either way, but this was the approach that required less responsibility on the allocator and less code in general.

MarcelKoch commented on cuda/base/stream.cpp:
```diff
@@ -30,58 +30,47 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-        auto cuda = gko::CudaExecutor::create(0, ref, true);
-    });
-}
+cuda_stream::cuda_stream() : stream_{}, device_id_{-1} {}
```
```suggestion
cuda_stream::cuda_stream() : stream_{nullptr}, device_id_{-1} {}
```
I'm not sure about the `-1`, since passing the nullptr as a stream is valid and results in the default stream getting used. So perhaps a valid device id would be appropriate.

upsj replied:
agree, the device_id doesn't play an important role except for the destruction, and we don't expose it otherwise

MarcelKoch commented on hip/base/roctx.hip.cpp:
```diff
@@ -0,0 +1,72 @@
+{
+    roctxRangePush(name);
+}
+
```
```suggestion


```

upsj commented on include/ginkgo/core/base/memory.hpp:
```diff
@@ -0,0 +1,215 @@
+
+    void deallocate(void* ptr) const override;
+
+    CudaAsyncAllocator(CUstream_st* stream);
```
should be possible, since we have a common base class.

upsj replied:
this is now supported by `check_environment`

upsj commented on include/ginkgo/core/base/memory.hpp:
```diff
@@ -0,0 +1,215 @@
+/*
+ * Allocator using hipMalloc.
+ */
+class HipAllocator : public HipAllocatorBase {
```
there are no suitable alternatives right now (with unified shared memory not really being available on AMD GPUs ATM), but there may be in the future.

MarcelKoch replied:
So they have the API for the different allocation modes, but they don't work? Weird.

upsj replied:
no, but I would assume they add at least USM allocation in the future

MarcelKoch replied:
ROCM 5.3 has managed allocation and host allocation (https://docs.amd.com/bundle/HIP_Supported_CUDA_API_Guide_v5.3/page/CUDA_Runtime_API_functions_supported_by_HIP.html#9_memory_management), or is that not what we are talking about here?

upsj replied:
I added the matching HIP allocators

upsj commented:
I agree with your suggestion to sync with application users, I'll see if I can find somebody with suitable experience. ATM, I have two possible interface variants in mind:
1. stateful allocators with `allocate(size_type)/deallocate(void*)` that store stream handle and maybe device ID to allow unified CPU/GPU allocation directly. Can lead to mismatches, which we may need to check for at executor construction time?
2. (by default) stateless allocators with `allocate(stream_handle, size_type)/deallocate(stream_handle, void*)` for GPUs and without stream_handle for CPU. This makes mismatches impossible, but makes it harder to use the same allocator for CPU and GPU.

ATM, I am leaning towards 1. i.e. adding `get_device_id` and `is_stream_ordered/get_stream` functions so that executors can check at construction time whether the allocator was set up correctly. What do you think @MarcelKoch?

MarcelKoch commented:
I was too subtle with my remark, sorry about that. I just meant that @pratikvn should take a look at it, because he mentioned that he has a use case for custom allocators.

MarcelKoch commented:
I think 1. is a suitable approach. But then we should add the checks to the executor constructor.
Another approach could be to have intermediate classes which have a function like `std::unique_ptr<Allocator> create_allocator(const Executor*)`, but that might complicate the class hierarchy a bit and thus make it harder for users to provide custom allocators.

upsj commented on core/device_hooks/hip_hooks.cpp:
```diff
@@ -53,12 +55,29 @@ version version_info::get_hip_version() noexcept
+    std::shared_ptr<HipAllocatorBase> alloc, GKO_HIP_STREAM_STRUCT* stream)
+{
+    return std::shared_ptr<HipExecutor>(
+        new HipExecutor(device_id, std::move(master),
+                        std::make_shared<HipAllocator>(), stream));
```
no, now that we have multiple supported allocators, we should use them.

upsj commented on cuda/base/memory.cpp:
```diff
@@ -0,0 +1,196 @@
+#else  // Fall back to regular allocation
+
+
+CudaAsyncAllocator::CudaAsyncAllocator(cudaStream_t stream) : stream_{} {}
```
true, that makes finding user-side bugs easier

upsj commented on include/ginkgo/core/base/memory.hpp:
```diff
@@ -0,0 +1,215 @@
+public:
+    DpcppAllocatorBase(sycl::queue* queue);
+
+    void* allocate(size_type num_bytes) const final;
+
+    void deallocate(void* ptr) const final;
```
only DPC++ has a queue that fully describes the device, for others, I need to set the environment from outside `allocate`, otherwise the interfaces would be more complex, and maybe do unnecessarily many cudaGetDevice calls. So I decided to only specialize the interface for DPC++.

upsj replied:
I removed the DPC++ allocator for now, since it requires more changes in the usage of queues.

upsj commented on core/test/base/executor.cpp:
```diff
@@ -310,35 +282,6 @@ TEST(HipExecutor, KnowsItsDeviceId)
-}
-
-
-TEST(HipExecutor, CanSetDefaultDeviceResetBoolean)
-{
-    auto omp = gko::OmpExecutor::create();
-    auto hip = gko::HipExecutor::create(0, omp, true);
-
-    ASSERT_EQ(true, hip->get_device_reset());
-}
```
if we no longer call DeviceReset, I think we should only return `false` to make it clearer to users. The deprecation warnings already do most of the job though.

MarcelKoch reviewed (approved):
LGTM, only one minor nit.

Regarding the user application, I think @pratikvn mentioned Amrex as a use case: https://amrex-codes.github.io/amrex/docs_html/GPU.html#memory-allocation. From the looks of it, it would match pretty well.

upsj commented on hip/test/base/memory.hip.cpp:
```diff
@@ -0,0 +1,126 @@
+    gko::array<int> data{exec_with_unified, {1, 2}};
+    exec->synchronize();
+
+    ASSERT_EQ(data.get_const_data()[0], 1);
```
I want to access the data from the host side, GKO_ASSERT_ARRAY_EQ would be using exec->copy instead

yhmtsai reviewed (commented):
how to call DeviceReset safely actually?
I assume we need to call it after all device-related objects are deleted. Thus, need another scope as following, is it right?

```
int main() {
{// using ginkgo including executor construction}
DeviceReset();
}
```
I also add @pratikvn into the reviewer list to ensure it fits the application needs.

pratikvn reviewed (commented):
Generally looks good to me. I definitely like this newer and nicer separation that is closer to my previous proposal of `MemorySpace`s :)

The way I see it, users can now create their own allocators, but need to derive from the `CudaAllocatorBase` for cuda, `HipAllocatorBase` for hip and override the `allocate` and `deallocate` functions. Is that right ? 

I think we need to make sure the device_id is always the same. I think in general this might not be the case for streams though. In many cases, it might be beneficial to use different streams for execution and memory operations (maintaining consistency between them). But I think in our current case, I think it is okay, as we always have only one stream that the executor takes.

upsj commented on cuda/base/memory.cpp:
```diff
@@ -0,0 +1,217 @@
+#else  // Fall back to regular allocation
+
+
+CudaAsyncAllocator::CudaAsyncAllocator(cudaStream_t stream) : stream_{stream} {}
```
good idea, construction only gets called once

upsj commented on cuda/base/memory.cpp:
```diff
@@ -0,0 +1,217 @@
+bool CudaAsyncAllocator::check_environment(int device_id,
+                                           CUstream_st* stream) const
+{
+    return stream == stream_;
```
the async allocator doesn't know the stream ID, and I don't see a way to query the device ID from the stream.

pratikvn replied:
I guess if the stream is correct, we can assume that the device_id is also correct ?

upsj replied:
yes, or at least if you try to use the stream with the wrong device, you get an error

upsj commented on include/ginkgo/core/base/executor.hpp:
```diff
@@ -1355,26 +1307,38 @@ class EnableDeviceReset {
      * @param device_reset  whether to allow a device reset or not
      */
-    void set_device_reset(bool device_reset) { device_reset_ = device_reset; }
+    [[deprecated(
+        "device_reset is no longer supported, call "
+        "cudaDeviceReset/hipDeviceReset manually")]] void
```
what do you have in mind? we have wrappers for the functions internally already

upsj replied:
On the question whether it needs to be used, I feel like most of the guidance is pretty old - I just checked again, the requirement to call cudaDeviceReset for `compute-sanitizer` has been removed in a recent version, the NSight Systems documentation is a bit unclear, especially what happens in the case of crashes?

> Nsight Systems will not have information about CUDA events that were still in device buffers when analysis terminated. It is a good idea, if using cudaProfilerAPI to control analysis to call cudaDeviceReset before ending analysis.

>  Normally, in order to collect full CUDA trace, the application needs to finalize the device used for CUDA work (call cudaDeviceReset(), and then let the application gracefully exit (as opposed to crashing).

Your suggestion may lead to some issues, since it requires the executor to exist until the `reset_device` function is called, but by calling it we basically destroy all allocations, so ideally we should be calling this outside the scope of the executor, like
```
int main() {
    {
        run_application();
    }
    cudaDeviceReset();
}
```

pratikvn replied:
It seems with both `ncu` and `nsys`, this branch (without the device reset) works the same as develop, so maybe they dont require a device_reset anymore ? 

With that, I guess most applications should not have a need to call reset anymore, so I guess it is okay that we dont provide that anymore.

upsj commented on cuda/base/executor.cpp:
```diff
@@ -65,25 +60,43 @@ namespace gko {
+    int device_id, std::shared_ptr<Executor> master,
+    std::shared_ptr<CudaAllocatorBase> alloc, cudaStream_t stream)
+{
+    if (!alloc->check_environment(device_id, stream)) {
+        throw Error{__FILE__, __LINE__,
+                    "Allocator uses incorrect stream or device ID."};
+    }
```
we only provide the device ID for allocators that work on both host and device, so this is not possible generally across all allocators.

upsj commented on cuda/test/base/cuda_executor.cu:
```diff
@@ -103,18 +104,19 @@ protected:
 #ifdef GKO_TEST_NONDEFAULT_STREAM
         cuda = gko::CudaExecutor::create(
-            0, omp, false, gko::default_cuda_alloc_mode, stream.get());
+            0, omp, std::make_shared<gko::CudaAllocator>(), stream.get());
```
I opened #1364 to handle this overlap between the types

upsj replied:
fixed

pratikvn reviewed (approved):
I had a look at what applications may need, with SUNDIALS as an example. It looks like this interface should be sufficient. 

These applications would need to create their own allocators deriving from `XAllocatorBase` for the different backends (`Cuda`, `Hip` etc) and pass that into executor object where they create the ginkgo executors. These allocators of course override the `allocate` and `deallocate` functions.
